### PR TITLE
split integration and unit tests

### PR DESCRIPTION
**Pull request by niemannd**
Well, the title says all. This splits the tests into unit tests (run on every build) and integration tests.
This should allow us to develop and run basic unit tests without the need for a meilisearch instance.

i'm not yet satisfied with that solution, maybe remove the test Classpaths. 

/discuss

### CODE REVIEW

**Comment by eskombro** [comment]
Hello @niemannd ! 

I'm not sure I understand something, when you say 

> This should allow us to develop and run basic unit tests without the need for a meilisearch instance

do you plan to use a mock server response for testing?

In the rest of SDKs development, we often use a MeiliSearch running instance for testing, and there's no separation between `unit` tests and `integration` tests, they are all the same. I think the process is fast and quite simple. I personally prefer to use a docker image for testing and the process is quite simple. When I'm doing a PR, I'm quite sure tests will succeed, as I did run the same tests locally just before.

So I'd be really happy to read what this approach will add for development / contribution from your point of view :)

**Comment by niemannd** [comment]
> do you plan to use a mock server response for testing?

I plan to mock class dependencies in unit tests.
This way we dont write classes that only work with a specific implementation of a dependency and rather just need a specific behaviour. 
Example: We want to test the Documents class. This class uses the MeiliSearchHttpRequest class. In our unit test we dont want to test the MeiliSearchHttpRequest class, just the behaviour. This behaviour is "method get(...) returns the response body as string". 
This way the mock would just need to return a string we define, nothing more. Makes testing specific cases much easier.

Unit tests basically describe what behaviour we expect from our class dependencies. I dont need to rely on 100% bug-free implementations of class dependencies. I dont need to rely on a live instance of meilisearch. Integration tests exist to test that our defined behaviour matches the real behaviour. 

A side effect of mocking most class dependencies is that your code becomes more modular because you have to design your classes to allow mocking. Look at the Documents class for example. To unit test this class, i would have to mock the MeiliSearchHttpRequest class. The way the class is written this is currently not possible.

**Comment by eskombro** [comment]
Thanks for the explanation @niemannd !

This would be really interesting, as you point out, for making tests more atomic and independent, and can make development easier, even if at the same time it introduces more work as the mocked behaviors sometimes need to be maintained/updated. 

It is really important, though, that we have a reliable integration set to check that everything is running smoothly for the whole SDK on the CI at integration time.

Great! :)

**Comment by eskombro** [comment]
I'm working on an integration tests PR https://github.com/meilisearch/meilisearch-java/pull/32, running docker on the CI. I'll try to merge it very soon, is it ok if you rebase your branch on it and split the tests?

**Comment by niemannd** [comment]
sure

**Comment by niemannd** [comment]
Regarding my last commit: I reworked the integration test setup a little bit, now using Tag Annotations provided by JUnit to seperate integration tests from the unit tests. This way we dont need a second SourceSet. 

I also learned that i shouldn't use the at symbol in commit messages. Sorry Tom :)

File: src/test/java/com/meilisearch/sdk/SearchRequestTest.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class SearchRequestTest {
+
+	@Test
+	void getQuery() {
+		SearchRequest classToTest = new SearchRequest("This is a Test");
```
Comment (eskombro) [inline src/test/java/com/meilisearch/sdk/SearchRequestTest.java:11]: I don't think that the default values should be handled by the SDK. I think that ideally, the query should be built only with the values that the user specifies, so in this case, a query "This is a test" shouldn't create the query `?q=This%20is%20a%20Test&offset=0&limit=20&attributesToRetrieve=*&cropLength=200&matches=false` but just `?q=This%20is%20a%20Test` and let the rest be handled by MeiliSearch itself. What do you think?

I know this is not an issue form these tests, but rather the SDK itself 👍 Just would like to have your opinion on it

**Review by eskombro — APPROVED** [review]
Thanks for this PR! 🎉 

Added a few comments/questions, but in the overall, it improves the testing flow and content, so quite glad to see this contribution!

File: .github/workflows/gradle.yml [hunk 2]
```diff
@@ -16,13 +16,13 @@ jobs:
 
         steps:
             -   uses: actions/checkout@v2
+            -   name: Docker setup
+                run: docker run -d -p 7700:7700 getmeili/meilisearch:latest ./meilisearch --no-analytics=true --no-sentry=true --master-key='masterKey'
```
[the remark this answers is not in the document]
Comment (niemannd) [reply .github/workflows/gradle.yml:20]: As i dont know what exactly triggers a sentry report, with the possible testing of the error handling and with the aspect that github actions is active for every fork i disabled sentry. A hard bug shoud be visible from the integration tests, but if you want sentry i have no problem with activating it.

File: build.gradle [hunk 3]
```diff
@@ -48,7 +56,18 @@ task buildJar(type: Jar) {
 }
 
 test {
-	useJUnitPlatform()
+	useJUnitPlatform {
```
[the remark this answers is not in the document]
Comment (niemannd) [reply build.gradle:59]: This config will exclude every Test or Class that is annotated with [at]Tag("integration") 
This allows us to also specify other tags, for example for extrem slow tests.

File: src/test/java/com/meilisearch/sdk/SearchRequestTest.java (hunk 1 above)
Comment (niemannd) [reply src/test/java/com/meilisearch/sdk/SearchRequestTest.java:11]: I just included this test as an example for a unit test :) The unit test folder looked a bit empty :D

File: .github/workflows/gradle.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (eskombro) [reply .github/workflows/gradle.yml:20]: No, it's ok, I just wanted to know if there was a specific reason for you to disable it, but it's totally fine if it isn't used in this context.

File: src/test/java/com/meilisearch/sdk/SearchRequestTest.java (hunk 1 above)
Comment (eskombro) [reply src/test/java/com/meilisearch/sdk/SearchRequestTest.java:11]: Indeed, quite a bit empty haha

**Pull request merged**